File: src/extras-kio/libjob.cpp
#include <QByteArray>
#include <QString>
#include <QUrl>
#include <job.h>
#include "libjob.h"
#include "libjob.hxx"

libqt_string KIO_BuildErrorString(int errorCode, const libqt_string errorText) {
    QString errorText_QString = QString::fromUtf8(errorText.data, errorText.len);
    auto _ret = KIO::buildErrorString(static_cast<int>(errorCode), errorText_QString);
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string KIO_RawErrorDetail(int errorCode, const libqt_string errorText, const QUrl* reqUrl, int method) {
    QString errorText_QString = QString::fromUtf8(errorText.data, errorText.len);
    QByteArray _qb = KIO::rawErrorDetail(static_cast<int>(errorCode), errorText_QString, reqUrl, static_cast<int>(method));
    libqt_string _str;
    _str.len = _qb.length();
    _str.data = static_cast<char*>(malloc(_str.len));
    memcpy((void*)_str.data, _qb.data(), _str.len);
    return _str;
}

File: src/extras-kio/libjob.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const QUrl = @import("libqt6").QUrl;
const std = @import("std");

/// ### [Upstream resources](https://api.kde.org/kio.html)
pub const KIO = extern struct {
    /// ### [Upstream resources](https://api.kde.org/kio.html)
    ///
    /// The pointer to the underlying Qt C++ object
    ///
    ptr: QtC.KIO,

    pub const _is_KIO = {};

    /// ### [Upstream resources](https://api.kde.org/kio.html#buildErrorString)
    ///
    /// ## Parameter(s):
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    /// ` errorCode: i32 `
    ///
    /// ` errorText: []const u8 `
    ///
    pub fn BuildErrorString(allocator: std.mem.Allocator, errorCode: i32, errorText: []const u8) []const u8 {
        const errorText_str = qtc.libqt_string{
            .len = errorText.len,
            .data = errorText.ptr,
        };
        var _str = qtc.KIO_BuildErrorString(@bitCast(errorCode), errorText_str);
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("kio.BuildErrorString: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// ### [Upstream resources](https://api.kde.org/kio.html#rawErrorDetail)
    ///
    /// ## Parameter(s):
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    /// ` errorCode: i32 `
    ///
    /// ` errorText: []const u8 `
    ///
    /// ` reqUrl: QUrl `
    ///
    /// ` method: i32 `
    ///
    pub fn RawErrorDetail(allocator: std.mem.Allocator, errorCode: i32, errorText: []const u8, reqUrl: anytype, method: i32) []u8 {
        const errorText_str = qtc.libqt_string{
            .len = errorText.len,
            .data = errorText.ptr,
        };
        comptime _ = @TypeOf(reqUrl)._is_QUrl;
        var _bytearray: qtc.libqt_string = qtc.KIO_RawErrorDetail(@bitCast(errorCode), errorText_str, @ptrCast(reqUrl.ptr), @bitCast(method));
        defer qtc.libqt_string_free(&_bytearray);
        const _ret = allocator.alloc(u8, _bytearray.len) catch @panic("kio.RawErrorDetail: Memory allocation failed");
        @memcpy(_ret, _bytearray.data[0.._bytearray.len]);
        return _ret;
    }
};
